### ISSUE TITLE
[Click-to-deploy] install istio

### ISSUE THREAD

**Issue by lluunn** [kubeflow/kubeflow#2052]
See https://github.com/kubeflow/website/pull/319.
It should be a simple kubectl apply.

@jlewi 
@kunmingg

**Comment by lluunn** [comment]
Our istio manifest is yaml not ksonnet because
1. it's large
1. we need to update it when istio has a new version
1. user won't need to change it after deployed, so no params need to be set

Then how do we install it with bootstrapper and save it with other manifest? I can think of several ways.
1. In ks_app, do `ks import` to drop in the yaml. But it will create a component per k8s resource, so there will be thousands of files which is not acceptable(?).
2. Use some hack to make it a ks component, e.g. a k8s job that deploys it.
3. save it as yaml, but not part of ks app, just like deployment manager config.

I think No.3 is better.
@kunmingg likes No.2 because it's more consistent with other components.
@jlewi thought?

**Comment by jlewi** [comment]
kfctl.sh already supports using raw YAML for some components. These are stored in the k8s_specs directory
https://www.kubeflow.org/docs/started/getting-started-gke/#app-layout

So does this work for ISTIO? Do we just need to add this support to click to deploy?

**Comment by lluunn** [comment]
Yes, this is what I meant by No.3

**Comment by kunmingg** [comment]
Can we apply through client library instead of running kubectl from bash?
Potential risk of bash kubectl:
Need to take care of credential isolation when running kubectl in parallel. We can not use default ```.kubeconfig```

**Comment by lluunn** [comment]
If each deployment creates its config, and we use kubectl with `--kubeconfig`, is this acceptable?

**Comment by jlewi** [comment]
Here's a link to some code using the REST client:
https://github.com/kubeflow/tf-operator/blob/e4a436da92e198dcb88c89c33010608e0c8a23bf/pkg/util/k8sutil/tf_job_client.go#L88

**Comment by lluunn** [comment]
IIUC, to use REST client, you need to set the api group/version, and the resource type in the request.
https://godoc.org/k8s.io/client-go/rest#Request

api group/version can be parsed from yaml file for each resource.
However, the resource type is the plural of the "Kind", which is not available by only looking at the yaml (I don't think there is a reliable kind-->pluralKind converter)
So I am not sure if REST client will help.

I found that client-go has dynamic client which can operate on arbitrary resource.
However, it's discouraged by the team to use it.
https://github.com/kubernetes/client-go/issues/394

**Comment by jlewi** [comment]
@lluunn What are the next steps?

**Comment by lluunn** [comment]
e2e test, and then enable it by default

**Comment by jlewi** [comment]
@lluunn Do we have a separate issue to install it as part of the no kfctl go binary?
The gocode is here:
https://github.com/kubeflow/kubeflow/tree/master/bootstrap/pkg

The new gobinary works by defining kfApps for different package installers (e.g. ksonnet, kustomize, etc..)
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/kfctl/README.md#kfapp-interface

We have a kfapp defined for ksonnet but not YAML or kustomize.

Since we are installing ISTIO from its YAML manifest; could we just treat it as Kustomize? I think we would just need to add a kustomize file and then combine that with the YAML file.

**Comment by bikramnehra** [comment]
@jlewi my understanding might be limited here, but is `helm` an option here:
https://istio.io/docs/setup/kubernetes/helm-install/#option-2-install-with-helm-and-tiller-via-helm-install

At SUSE we rely on `helm` for most components including Istio:  
https://github.com/SUSE/scf/blob/develop/make/istio/run#L22

**Comment by lluunn** [comment]
helm doesn't provide the [customization](https://github.com/kubeflow/kubeflow/blob/master/dependencies/istio/install/README.md) we want.

We will have to make upstream PR and wait for their release every time we need to customize

**Comment by jlewi** [comment]
@lluunn Can we close this issue?

**Comment by lluunn** [comment]
C2D currently doesn't have the user visible option to install KF with istio on, although the functionality is there.

I think we can this open until we turn on Istio by default?

**Comment by lluunn** [comment]
will use #3007 to finish this
